src/Microsoft.FeatureManagement/FeatureFilters/TimeWindowFilter.cs

```diff
@@ -56,7 +68,74 @@ public Task<bool> EvaluateAsync(FeatureFilterEvaluationContext context)
                 return Task.FromResult(false);
             }
 
-            return Task.FromResult((!settings.Start.HasValue || now >= settings.Start.Value) && (!settings.End.HasValue || now < settings.End.Value));
+            //
+            // Hit the first occurrence of the time window
+            if ((!settings.Start.HasValue || now >= settings.Start.Value) && (!settings.End.HasValue || now < settings.End.Value))
+            {
+                return Task.FromResult(true);
+            }
+
+            if (settings.Recurrence != null)
+            {
+                //
+                // The reference of the object will be used for hash key.
+                // If there is no pre-bounded settings attached to the context, there will be no cached filter settings and each call will have a unique settings object.
+                // In this case, the cache for recurrence settings won't work.
+                if (context.Settings != null)
+                {
+                    DateTimeOffset cachedTime = _recurrenceCache.GetOrAdd(
```

Comment:
this cache grows unbound.

Reply:
We may want to use a memory cache to support expiration.

Also, I expect the cache key should be the feature name. Otherwise, tweaks to the settings for a given feature would result in a new cache item and a leftover cache item that's useless.

Reply:
We have `AddMemoryCache` call within the `AddFeatureManagement`. While using DI, feature manager will use the shared IMemoryCache. Should TimeWindowFilter follow this pattern? Or it should have its own memory cache?

Reply:
> I expect the cache key should be the feature name.

I remember we discussed this before. A feature flag can have multiple timewindow filters with recurrence settings. 

> tweaks to the settings for a given feature would result in a new cache item

This is expected, isn't it? If we use feature name as cache key, when recurrence setting of a feature is changed, how we know the old cache with the key of the feature name should be discarded?

> a leftover cache item that's useless.

This is a problem. I agree that we should use Memory cache and set an expiration time.

Reply:
the key should be the flag name, and the cached value should have something to indicate that it's still relevant.

Reply:
> A feature flag can have multiple timewindow filters with different recurrence settings. If we want to use feature name as cache key, we will need additional information about the filter index.

Current `FeatureFIlterEvaluationContext` doesn't include such information.

Reply:
Discussed offline. Will still use settings reference as cache key and use IMemoryCache instead of ConcurrentDictionary.